// Copyright 2022 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

syntax = "proto3";

package google.scp.core.transaction_manager.proto;

import "scp/cc/core/common/proto/common.proto";

// For faster allocations of sub-messages.
option cc_enable_arenas = true;

enum TransactionLogType {
  TRANSACTION_LOG_TYPE_UNKNOWN = 0;
  TRANSACTION_LOG = 1;
  TRANSACTION_PHASE_LOG = 2;
};

enum TransactionPhase {
  TRANSACTION_PHASE_TYPE_UNKNOWN = 0;
  NOT_STARTED = 1;
  BEGIN = 2;
  PREPARE = 3;
  COMMIT = 4;
  COMMIT_NOTIFY = 5;
  COMMITTED = 6;
  ABORT_NOTIFY = 7;
  ABORTED = 8;
  END = 9;
};

message TransactionCommandLog_1_0 {
  bytes command_body = 1;
};

message TransactionPhaseLog_1_0 {
  core.common.proto.Uuid id = 1;
  TransactionPhase phase = 2;
  bool failed = 3;
  core.common.proto.ExecutionResult result = 4;
};

message TransactionLog_1_0 {
  core.common.proto.Uuid id = 1;
  uint64 timeout = 2;
  bool is_coordinated_remotely = 3;
  string transaction_secret = 4;
  string transaction_origin = 5;
  repeated TransactionCommandLog_1_0 commands = 6;
};

message TransactionEngineLog_1_0 {
  TransactionLogType type = 1;
  bytes log_body = 2;
};

message TransactionEngineLog {
  core.common.proto.Version version = 1;
  bytes log_body = 2;
};
